include/gatekeeper_varip.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_VARIP_H_
+#define _GATEKEEPER_VARIP_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/*
+ * The code of this file is mostly a copy of the Linux kernel.
+ */
+
+/* TODO Add support for variable IP headers for all the functional blocks. */
+
+/* NextHeader field of IPv6 header. */
+
+/* Hop-by-hop option header. */
+#define NEXTHDR_HOP      (0)
+
+/* Routing header. */
+#define NEXTHDR_ROUTING  (43)
+
+/* Fragmentation/reassembly header. */
+#define NEXTHDR_FRAGMENT (44)
+
+/* Authentication header. */
+#define NEXTHDR_AUTH     (51)
+
+/* No next header. */
+#define NEXTHDR_NONE     (59)
+
+/* Destination options header. */
+#define NEXTHDR_DEST     (60)
+
+#define NEXTHDR_MAX      (255)
+
+bool ipv6_ext_hdr(uint8_t nexthdr);
+
+/*
+ * Skip any extension headers.
+ *
+ * This function parses (probably truncated) exthdr set "hdr".
+ * "nexthdrp" initially points to some place,
+ * where type of the first header can be found.
+ *
+ * It skips all well-known exthdrs, and returns pointer to the start
+ * of unparsable area i.e. the first header with unknown type.
+ * If it is not NULL *nexthdr is updated by type/protocol of this header.
+ *
+ * NOTES: - if packet terminated with NEXTHDR_NONE it returns NULL.
+ *        - it may return pointer pointing beyond end of packet,
+ *	    if the last recognized header is truncated in the middle.
+ *        - if packet is truncated, so that all parsed headers are skipped,
+ *	    it returns NULL.
+ *	  - First fragment header is skipped, not-first ones
+ *	    are considered as unparsable.
```

Comment:
The code just parses the first fragment header. Notice that, only when it's the first fragment (i.e., *frag_offp == 0), ``if (ntohs(*frag_offp) & ~0x7)`` is false. If there is more than one fragment header, then the above condition becomes true, and the function considers this as unparsable.